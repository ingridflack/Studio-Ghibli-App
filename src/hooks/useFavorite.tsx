import {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState,
} from "react";
import { toast } from "react-toastify";
import { IMovie } from "../types";

interface FavoriteProviderProps {
  children: ReactNode;
}

interface FavoriteContextData {
  favorites?: IMovie[];
  checkFavorite: (movie: IMovie) => boolean;
  toggleFavorite: (movie: IMovie) => void;
}

const FavoriteContext = createContext<FavoriteContextData>(
  {} as FavoriteContextData
);

const STORAGE_FAVORITE_KEY = "@Ghibli:favorites";

export function FavoriteProvider({
  children,
}: FavoriteProviderProps): JSX.Element {
  const [favorites, setFavorites] = useState<IMovie[] | undefined>();

  useEffect(() => {
    const storagedFavorites = localStorage.getItem(STORAGE_FAVORITE_KEY);
    const favorites = storagedFavorites ? JSON.parse(storagedFavorites) : [];

    setFavorites(favorites);
  }, []);

  useEffect(() => {
    if (favorites) {
      localStorage.setItem(STORAGE_FAVORITE_KEY, JSON.stringify(favorites));
    }
  }, [favorites]);

  const checkFavorite = (movie: IMovie): boolean =>
    !!favorites?.some((item) => item.id === movie.id);

  const addFavorite = (movie: IMovie) => {
    setFavorites((favorites) => (favorites ? [...favorites, movie] : []));

    toast.success("Movie added to favorites.", {
      position: toast.POSITION.BOTTOM_RIGHT,
    });
  };

  const removeFavorite = (movie: IMovie) => {
    setFavorites((favorites) =>
      favorites?.filter((item) => item.id !== movie.id)
    );

    toast.success("Movie removed from favorites.", {
      position: toast.POSITION.BOTTOM_RIGHT,
    });
  };

  const toggleFavorite = (movie: IMovie) => {
    if (checkFavorite(movie)) {
      removeFavorite(movie);
    } else {
      addFavorite(movie);
    }
  };

  return (
    <FavoriteContext.Provider
      value={{
        favorites,
        checkFavorite,
        toggleFavorite,
      }}
    >
      {children}
    </FavoriteContext.Provider>
  );
}

export function useFavorite() {
  const context = useContext(FavoriteContext);

  return context;
}
