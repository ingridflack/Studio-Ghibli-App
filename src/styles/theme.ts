const theme = {
  colors: {
    background: "#000",

    gray: {
      white: "#f6f6f6",
      platinum: "#e1e1e6",
      silver: "#a8a8b3",
      lightGray: "#7b8c98",
      dark: "#141414",
    },

    primary: "#e50914",
    secondary: "#b20710",
  },
  fontSize: {
    tiny: "0.65rem",
    small: "0.75rem",
    medium: "0.875rem",
    normal: "1rem",
    large: "1.5rem",
  },
} as const;

export type Theme = typeof theme;

export default theme;
