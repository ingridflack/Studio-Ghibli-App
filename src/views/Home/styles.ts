import styled, { css } from "styled-components";

export const MoviesList = styled.div`
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  gap: 2rem;

  @media screen and (max-width: 768px) {
    grid-template-columns: 1fr 1fr;
    gap: 1rem;
  }

  @media screen and (max-width: 576px) {
    grid-template-columns: 1fr;
  }
`;

export const Input = styled.input`
  border-radius: 0.75rem;
  padding: 0.5rem 0.5rem 0.5rem 2.5rem;
  border: 2px solid transparent;
  outline: none;
  transition: border 0.3s ease;
  position: relative;
  width: 100%;

  &:focus {
    border: 2px solid ${({ theme }) => theme.colors.primary};
  }
`;

export const Form = styled.form`
  ${({ theme }) => css`
    grid-area: form;
    position: relative;

    svg {
      color: ${theme.colors.primary};
      font-size: ${theme.fontSize.normal};
      position: absolute;
      top: 50%;
      left: 1rem;
      transform: translateY(-50%);
    }
  `}
`;
