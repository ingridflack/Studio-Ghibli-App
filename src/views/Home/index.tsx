import { ChangeEvent, useMemo, useState } from "react";

import MovieCard from "../../components/MovieCard";
import Header from "../../components/Header";
import { IMovie } from "../../types";

import * as S from "./styles";
import Alert from "../../components/Alert";
import { AiOutlineSearch } from "react-icons/ai";
import { MainContainer } from "../../styles/shared";
interface HomeProps {
  movies: IMovie[];
}

const Home = ({ movies }: HomeProps) => {
  const [searchValue, setSearchValue] = useState("");

  const handleSearch = (e: ChangeEvent<HTMLInputElement>) => {
    const search = e.target.value;

    setSearchValue(search);
  };

  const filteredMovies = useMemo(
    () =>
      movies.filter(({ title }) =>
        title.toLowerCase().includes(searchValue.toLowerCase())
      ),
    [movies, searchValue]
  );

  return (
    <>
      <Header />

      <S.Form>
        <S.Input
          placeholder="Search"
          value={searchValue}
          onChange={handleSearch}
        />
        <AiOutlineSearch />
      </S.Form>

      <S.MoviesList>
        {filteredMovies.map((movie) => (
          <MovieCard key={movie.id} movie={movie} />
        ))}

        {!filteredMovies.length && <Alert>No results found</Alert>}
      </S.MoviesList>
    </>
  );
};

export default Home;
