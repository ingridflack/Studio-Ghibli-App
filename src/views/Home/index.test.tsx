import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import Home from ".";
import theme from "../../styles/theme";

const useRouter = jest.spyOn(require("next/router"), "useRouter");
const useFavorite = jest.spyOn(
  require("../../hooks/useFavorite"),
  "useFavorite"
);

const MOVIE_LIST_MOCK = [
  {
    id: "234",
    title: "Title test",
    description: "Description test",
    director: "Director test",
    producer: "Producer test",
    movie_banner: "/assets/studio-ghibli-logo.webp",
    rt_score: "90",
    release_date: "1980",
    people: ["Person 1"],
    original_title: "Original title test",
  },
  {
    id: "2345",
    title: "Title test 2",
    description: "Description test 2",
    director: "Director test 2",
    producer: "Producer test 2",
    movie_banner: "/assets/studio-ghibli-logo.webp",
    rt_score: "90 2",
    release_date: "1980 2",
    people: ["Person 12"],
    original_title: "Original title test 2",
  },
];

describe("Home", () => {
  useRouter.mockImplementation(() => ({
    pathname: "/",
  }));

  it("should render the movie list correctly", () => {
    (useFavorite as jest.Mock).mockReturnValue({
      checkFavorite: jest.fn().mockReturnValue(true),
    });

    const { getAllByTestId } = render(
      <ThemeProvider theme={theme}>
        <Home movies={MOVIE_LIST_MOCK} />
      </ThemeProvider>
    );

    const movieCard = getAllByTestId("movie-card-container");

    expect(movieCard).toHaveLength(2);
  });

  it("should match snapshot", () => {
    (useFavorite as jest.Mock).mockReturnValue({
      checkFavorite: jest.fn().mockReturnValue(true),
    });

    const { container } = render(
      <ThemeProvider theme={theme}>
        <Home movies={MOVIE_LIST_MOCK} />
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
