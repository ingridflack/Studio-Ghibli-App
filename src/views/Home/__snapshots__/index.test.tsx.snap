// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match snapshot 1`] = `
<div>
  <div
    class="sc-acc38eb5-0 jBEmtF"
  >
    <img
      alt="Studio Ghibli logo"
      class="sc-acc38eb5-1 hxYLtv"
      height="72"
      src="/assets/studio-ghibli-logo.webp"
      width="128"
    />
    <nav>
      <a
        class="sc-acc38eb5-2 iwGSMA"
        href="/"
      >
        Home
      </a>
      <a
        class="sc-acc38eb5-2 bYjima"
        href="/favorites"
      >
        Favorites
      </a>
    </nav>
  </div>
  <div
    class="sc-ac3e41b8-2 cewkpQ"
  >
    <input
      class="sc-ac3e41b8-1 fVEcxk"
      data-cy="input"
      placeholder="Search"
      value=""
    />
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0 0 11.6 0l43.6-43.5a8.2 8.2 0 0 0 0-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
      />
    </svg>
  </div>
  <div
    class="sc-ac3e41b8-0 kpVdd"
  >
    <a
      class="sc-74a0e932-0 jKUBJa"
      data-testid="movie-card-container"
      href="/movies/234"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 53.333333333333336% 0px 0px 0px;"
        />
        <img
          alt="Title test movie banner"
          class="sc-74a0e932-1 fqsWUy"
          data-nimg="responsive"
          data-testid="movie-banner"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
      <div
        class="sc-74a0e932-2 cpILKX"
      >
        <span
          class="sc-74a0e932-3 eWneaZ"
        >
          <span
            class="sc-74a0e932-4 dbGZIV"
            data-testid="card-title"
          >
            Title test
          </span>
          <div
            class="sc-ade4b8be-1 kbXxyf"
          >
            <div
              class="sc-ade4b8be-0 jUqyaW"
            >
              Remove from favorites
            </div>
            <button
              class="sc-74a0e932-6 dplNRY"
              data-testid="favorite-btn"
              type="button"
            >
              <svg
                data-testid="favorited"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M923 283.6a260.04 260.04 0 0 0-56.9-82.8 264.4 264.4 0 0 0-84-55.5A265.34 265.34 0 0 0 679.7 125c-49.3 0-97.4 13.5-139.2 39-10 6.1-19.5 12.8-28.5 20.1-9-7.3-18.5-14-28.5-20.1-41.8-25.5-89.9-39-139.2-39-35.5 0-69.9 6.8-102.4 20.3-31.4 13-59.7 31.7-84 55.5a258.44 258.44 0 0 0-56.9 82.8c-13.9 32.3-21 66.6-21 101.9 0 33.3 6.8 68 20.3 103.3 11.3 29.5 27.5 60.1 48.2 91 32.8 48.9 77.9 99.9 133.9 151.6 92.8 85.7 184.7 144.9 188.6 147.3l23.7 15.2c10.5 6.7 24 6.7 34.5 0l23.7-15.2c3.9-2.5 95.7-61.6 188.6-147.3 56-51.7 101.1-102.7 133.9-151.6 20.7-30.9 37-61.5 48.2-91 13.5-35.3 20.3-70 20.3-103.3.1-35.3-7-69.6-20.9-101.9z"
                />
              </svg>
            </button>
          </div>
        </span>
        <span
          class="sc-74a0e932-5 jZSNfa"
        >
          Original title test
        </span>
        <span
          class="sc-74a0e932-7 jWDmqT"
        >
          <strong>
            Director:
          </strong>
           
          Director test
        </span>
        <p
          class="sc-74a0e932-8 cogrSB"
        >
          Description test
        </p>
      </div>
    </a>
    <a
      class="sc-74a0e932-0 jKUBJa"
      data-testid="movie-card-container"
      href="/movies/2345"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 53.333333333333336% 0px 0px 0px;"
        />
        <img
          alt="Title test 2 movie banner"
          class="sc-74a0e932-1 fqsWUy"
          data-nimg="responsive"
          data-testid="movie-banner"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
      <div
        class="sc-74a0e932-2 cpILKX"
      >
        <span
          class="sc-74a0e932-3 eWneaZ"
        >
          <span
            class="sc-74a0e932-4 dbGZIV"
            data-testid="card-title"
          >
            Title test 2
          </span>
          <div
            class="sc-ade4b8be-1 kbXxyf"
          >
            <div
              class="sc-ade4b8be-0 jUqyaW"
            >
              Remove from favorites
            </div>
            <button
              class="sc-74a0e932-6 dplNRY"
              data-testid="favorite-btn"
              type="button"
            >
              <svg
                data-testid="favorited"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M923 283.6a260.04 260.04 0 0 0-56.9-82.8 264.4 264.4 0 0 0-84-55.5A265.34 265.34 0 0 0 679.7 125c-49.3 0-97.4 13.5-139.2 39-10 6.1-19.5 12.8-28.5 20.1-9-7.3-18.5-14-28.5-20.1-41.8-25.5-89.9-39-139.2-39-35.5 0-69.9 6.8-102.4 20.3-31.4 13-59.7 31.7-84 55.5a258.44 258.44 0 0 0-56.9 82.8c-13.9 32.3-21 66.6-21 101.9 0 33.3 6.8 68 20.3 103.3 11.3 29.5 27.5 60.1 48.2 91 32.8 48.9 77.9 99.9 133.9 151.6 92.8 85.7 184.7 144.9 188.6 147.3l23.7 15.2c10.5 6.7 24 6.7 34.5 0l23.7-15.2c3.9-2.5 95.7-61.6 188.6-147.3 56-51.7 101.1-102.7 133.9-151.6 20.7-30.9 37-61.5 48.2-91 13.5-35.3 20.3-70 20.3-103.3.1-35.3-7-69.6-20.9-101.9z"
                />
              </svg>
            </button>
          </div>
        </span>
        <span
          class="sc-74a0e932-5 jZSNfa"
        >
          Original title test 2
        </span>
        <span
          class="sc-74a0e932-7 jWDmqT"
        >
          <strong>
            Director:
          </strong>
           
          Director test 2
        </span>
        <p
          class="sc-74a0e932-8 cogrSB"
        >
          Description test 2
        </p>
      </div>
    </a>
  </div>
</div>
`;
